File: zig/src/indicators/marc_chaikin/advance_decline/advance_decline.zig
const std = @import("std");
const math = std.math;


const entities = @import("entities");
const Bar = entities.Bar;
const Quote = entities.Quote;
const Trade = entities.Trade;
const Scalar = entities.Scalar;
const quote_component = entities.quote_component;
const trade_component = entities.trade_component;
const indicator_mod = @import("../../core/indicator.zig");
const line_indicator_mod = @import("../../core/line_indicator.zig");
const build_metadata_mod = @import("../../core/build_metadata.zig");
const identifier_mod = @import("../../core/identifier.zig");
const metadata_mod = @import("../../core/metadata.zig");

const OutputArray = indicator_mod.OutputArray;
const LineIndicator = line_indicator_mod.LineIndicator;
const Identifier = identifier_mod.Identifier;
const Metadata = metadata_mod.Metadata;

/// Enumerates the outputs of the Advance-Decline indicator.
pub const AdvanceDeclineOutput = enum(u8) {
    /// The scalar value of the A/D line.
    value = 1,
};

/// Marc Chaikin's Advance-Decline (A/D) Line.
///
/// The Accumulation/Distribution Line is a cumulative indicator that uses volume
/// and price to assess whether a stock is being accumulated or distributed.
///
/// CLV = ((Close - Low) - (High - Close)) / (High - Low)
/// AD  = AD_previous + CLV × Volume
///
/// When High equals Low, the A/D value is unchanged (no division by zero).
pub const AdvanceDecline = struct {
    line: LineIndicator,
    ad: f64,
    value: f64,
    primed: bool,

    const mnemonic_str = "ad";
    const description_str = "Advance-Decline";

    pub fn init() AdvanceDecline {
        return .{
            .line = LineIndicator.new(
                mnemonic_str,
                description_str,
                null,
                null,
                null,
            ),
            .ad = 0,
            .value = math.nan(f64),
            .primed = false,
        };
    }

    pub fn deinit(_: *AdvanceDecline) void {}

    pub fn fixSlices(_: *AdvanceDecline) void {}

    /// Core update logic. For scalar/quote/trade, H=L=C so AD is unchanged.
    pub fn update(self: *AdvanceDecline, sample: f64) f64 {
        if (math.isNan(sample)) {
            return math.nan(f64);
        }
        return self.updateHlcv(sample, sample, sample, 1);
    }

    /// Updates the indicator with the given high, low, close, and volume values.
    pub fn updateHlcv(self: *AdvanceDecline, high: f64, low: f64, close: f64, volume: f64) f64 {
        if (math.isNan(high) or math.isNan(low) or math.isNan(close) or math.isNan(volume)) {
            return math.nan(f64);
        }

        const temp = high - low;
        if (temp > 0) {
            self.ad += ((close - low) - (high - close)) / temp * volume;
        }

        self.value = self.ad;
        self.primed = true;

        return self.value;
    }

    pub fn isPrimed(self: *const AdvanceDecline) bool {
        return self.primed;
    }

    pub fn getMetadata(self: *const AdvanceDecline, out: *Metadata) void {
        build_metadata_mod.buildMetadata(
            out,
            .advance_decline,
            self.line.mnemonic,
            self.line.description,
            &[_]build_metadata_mod.OutputText{
                .{ .mnemonic = self.line.mnemonic, .description = self.line.description },
            },
        );
    }

    pub fn updateScalar(self: *AdvanceDecline, sample: *const Scalar) OutputArray {
        const value = self.update(sample.value);
        return LineIndicator.wrapScalar(sample.time, value);
    }

    /// Shadows LineIndicator.updateBar to extract HLCV directly from the bar.
    pub fn updateBar(self: *AdvanceDecline, sample: *const Bar) OutputArray {
        const value = self.updateHlcv(sample.high, sample.low, sample.close, sample.volume);
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateQuote(self: *AdvanceDecline, sample: *const Quote) OutputArray {
        const value = self.update(self.line.extractQuote(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateTrade(self: *AdvanceDecline, sample: *const Trade) OutputArray {
        const value = self.update(self.line.extractTrade(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn indicator(self: *AdvanceDecline) indicator_mod.Indicator {
        return .{
            .ptr = @ptrCast(self),
            .vtable = &vtable,
        };
    }

    const vtable = indicator_mod.Indicator.VTable{
        .isPrimed = vtableIsPrimed,
        .metadata = vtableMetadata,
        .updateScalar = vtableUpdateScalar,
        .updateBar = vtableUpdateBar,
        .updateQuote = vtableUpdateQuote,
        .updateTrade = vtableUpdateTrade,
    };

    fn vtableIsPrimed(ptr: *anyopaque) bool {
        const self: *AdvanceDecline = @ptrCast(@alignCast(ptr));
        return self.isPrimed();
    }

    fn vtableMetadata(ptr: *anyopaque, out: *Metadata) void {
        const self: *const AdvanceDecline = @ptrCast(@alignCast(ptr));
        self.getMetadata(out);
    }

    fn vtableUpdateScalar(ptr: *anyopaque, sample: *const Scalar) OutputArray {
        const self: *AdvanceDecline = @ptrCast(@alignCast(ptr));
        return self.updateScalar(sample);
    }

    fn vtableUpdateBar(ptr: *anyopaque, sample: *const Bar) OutputArray {
        const self: *AdvanceDecline = @ptrCast(@alignCast(ptr));
        return self.updateBar(sample);
    }

    fn vtableUpdateQuote(ptr: *anyopaque, sample: *const Quote) OutputArray {
        const self: *AdvanceDecline = @ptrCast(@alignCast(ptr));
        return self.updateQuote(sample);
    }

    fn vtableUpdateTrade(ptr: *anyopaque, sample: *const Trade) OutputArray {
        const self: *AdvanceDecline = @ptrCast(@alignCast(ptr));
        return self.updateTrade(sample);
    }
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

const testing = std.testing;
const testdata = @import("testdata.zig");


fn roundTo(v: f64, comptime digits: comptime_int) f64 {
    const p = comptime blk: {
        var result: f64 = 1.0;
        for (0..digits) |_| result *= 10.0;
        break :blk result;
    };
    return @round(v * p) / p;
}

// High test data, 252 entries.
// Low test data, 252 entries.
// Close test data, 252 entries.
// Volume test data, 252 entries.
// Expected AD output, 252 entries.
test "AdvanceDecline with volume" {
    const highs = testdata.testHighs();
    const lows = testdata.testLows();
    const closes = testdata.testCloses();
    const volumes = testdata.testVolumes();
    const expected = testdata.testExpectedAD();

    var ad = AdvanceDecline.init();

    for (0..252) |i| {
        const v = ad.updateHlcv(highs[i], lows[i], closes[i], volumes[i]);
        try testing.expect(!math.isNan(v));
        try testing.expect(ad.isPrimed());

        const got = roundTo(v, 2);
        const exp = roundTo(expected[i], 2);
        try testing.expectEqual(exp, got);
    }
}

test "AdvanceDecline TA-Lib spot checks" {
    const highs = testdata.testHighs();
    const lows = testdata.testLows();
    const closes = testdata.testCloses();
    const volumes = testdata.testVolumes();

    var ad = AdvanceDecline.init();
    var values: [252]f64 = undefined;

    for (0..252) |i| {
        values[i] = ad.updateHlcv(highs[i], lows[i], closes[i], volumes[i]);
    }

    try testing.expectEqual(roundTo(-1631000.00, 2), roundTo(values[0], 2));
    try testing.expectEqual(roundTo(2974412.02, 2), roundTo(values[1], 2));
    try testing.expectEqual(roundTo(8707691.07, 2), roundTo(values[250], 2));
    try testing.expectEqual(roundTo(8328944.54, 2), roundTo(values[251], 2));
}

test "AdvanceDecline updateBar" {
    var ad = AdvanceDecline.init();

    const highs = testdata.testHighs();
    const lows = testdata.testLows();
    const closes = testdata.testCloses();
    const volumes = testdata.testVolumes();
    const expected = testdata.testExpectedAD();

    for (0..10) |i| {
        const bar = Bar{
            .time = 0,
            .open = highs[i],
            .high = highs[i],
            .low = lows[i],
            .close = closes[i],
            .volume = volumes[i],
        };

        const output = ad.updateBar(&bar);
        const scalar_val = output.slice()[0].scalar.value;

        const got = roundTo(scalar_val, 2);
        const exp = roundTo(expected[i], 2);
        try testing.expectEqual(exp, got);
    }
}

test "AdvanceDecline scalar update" {
    var ad = AdvanceDecline.init();

    // Scalar update: H=L=C, so range=0, AD unchanged (remains 0 after primed).
    const v = ad.update(100.0);
    try testing.expectEqual(@as(f64, 0), v);
    try testing.expect(ad.isPrimed());
}

test "AdvanceDecline NaN" {
    var ad = AdvanceDecline.init();

    try testing.expect(math.isNan(ad.update(math.nan(f64))));
    try testing.expect(math.isNan(ad.updateHlcv(math.nan(f64), 1, 2, 3)));
    try testing.expect(math.isNan(ad.updateHlcv(1, math.nan(f64), 2, 3)));
    try testing.expect(math.isNan(ad.updateHlcv(1, 2, math.nan(f64), 3)));
    try testing.expect(math.isNan(ad.updateHlcv(1, 2, 3, math.nan(f64))));
}

test "AdvanceDecline not primed initially" {
    var ad = AdvanceDecline.init();

    try testing.expect(!ad.isPrimed());
    try testing.expect(math.isNan(ad.update(math.nan(f64))));
    try testing.expect(!ad.isPrimed());
}

test "AdvanceDecline metadata" {
    var ad = AdvanceDecline.init();
    var meta: Metadata = undefined;
    ad.getMetadata(&meta);

    try testing.expectEqual(Identifier.advance_decline, meta.identifier);
    try testing.expectEqualStrings("ad", meta.mnemonic);
    try testing.expectEqualStrings("Advance-Decline", meta.description);
    try testing.expectEqual(@as(usize, 1), meta.outputs_len);
}

File: rs/src/indicators/marc_chaikin/advance_decline/advance_decline.rs
use crate::entities::bar::Bar;
use crate::entities::quote::Quote;
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::{Indicator, Output};
use crate::indicators::core::metadata::Metadata;

// ---------------------------------------------------------------------------
// Params
// ---------------------------------------------------------------------------

/// Parameters for the Advance-Decline indicator.
/// Advance-Decline requires HLCV bar data and has no configurable parameters.
pub struct AdvanceDeclineParams;

impl Default for AdvanceDeclineParams {
    fn default() -> Self {
        Self
    }
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

/// Enumerates the outputs of the Advance-Decline indicator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum AdvanceDeclineOutput {
    /// The scalar value of the advance-decline line.
    Value = 1,
}

// ---------------------------------------------------------------------------
// Indicator
// ---------------------------------------------------------------------------

const AD_MNEMONIC: &str = "ad";
const AD_DESCRIPTION: &str = "Advance-Decline";

/// Marc Chaikin's Advance-Decline (A/D) Line.
///
/// The Accumulation/Distribution Line is a cumulative indicator that uses volume
/// and price to assess whether a stock is being accumulated or distributed.
///
/// The value is calculated as:
///
///   CLV = ((Close - Low) - (High - Close)) / (High - Low)
///   AD  = AD_previous + CLV × Volume
///
/// When High equals Low, the A/D value is unchanged (no division by zero).
pub struct AdvanceDecline {
    ad: f64,
    value: f64,
    primed: bool,
}

impl AdvanceDecline {
    /// Creates a new Advance-Decline indicator.
    pub fn new(_params: &AdvanceDeclineParams) -> Result<Self, String> {
        Ok(Self {
            ad: 0.0,
            value: f64::NAN,
            primed: false,
        })
    }

    /// Updates the indicator with high, low, close, and volume values.
    pub fn update_hlcv(&mut self, high: f64, low: f64, close: f64, volume: f64) -> f64 {
        if high.is_nan() || low.is_nan() || close.is_nan() || volume.is_nan() {
            return f64::NAN;
        }

        let temp = high - low;
        if temp > 0.0 {
            self.ad += ((close - low) - (high - close)) / temp * volume;
        }

        self.value = self.ad;
        self.primed = true;

        self.value
    }

    /// Updates using a single sample value (H=L=C, so range=0, AD unchanged).
    pub fn update(&mut self, sample: f64) -> f64 {
        if sample.is_nan() {
            return f64::NAN;
        }

        self.update_hlcv(sample, sample, sample, 1.0)
    }
}

impl Indicator for AdvanceDecline {
    fn is_primed(&self) -> bool {
        self.primed
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::AdvanceDecline,
            AD_MNEMONIC,
            AD_DESCRIPTION,
            &[OutputText {
                mnemonic: AD_MNEMONIC.to_string(),
                description: AD_DESCRIPTION.to_string(),
            }],
        )
    }

    fn update_scalar(&mut self, sample: &Scalar) -> Output {
        let v = sample.value;
        let result = self.update_hlcv(v, v, v, 1.0);
        vec![Box::new(Scalar::new(sample.time, result))]
    }

    fn update_bar(&mut self, sample: &Bar) -> Output {
        let result = self.update_hlcv(sample.high, sample.low, sample.close, sample.volume);
        vec![Box::new(Scalar::new(sample.time, result))]
    }

    fn update_quote(&mut self, sample: &Quote) -> Output {
        let v = (sample.bid_price + sample.ask_price) / 2.0;
        let result = self.update_hlcv(v, v, v, 1.0);
        vec![Box::new(Scalar::new(sample.time, result))]
    }

    fn update_trade(&mut self, sample: &Trade) -> Output {
        let v = sample.price;
        let result = self.update_hlcv(v, v, v, 1.0);
        vec![Box::new(Scalar::new(sample.time, result))]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::testdata::testdata;
    fn round_to(v: f64, digits: u32) -> f64 {
        let p = 10f64.powi(digits as i32);
        (v * p).round() / p
    }

    #[test]
    fn test_advance_decline_with_volume() {
        let digits = 2;
        let highs = testdata::test_highs();
        let lows = testdata::test_lows();
        let closes = testdata::test_closes();
        let volumes = testdata::test_volumes();
        let expected = testdata::test_expected_ad();
        let count = highs.len();

        let mut ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();

        for i in 0..count {
            let v = ad.update_hlcv(highs[i], lows[i], closes[i], volumes[i]);
            assert!(!v.is_nan(), "[{}] expected non-NaN, got NaN", i);
            assert!(ad.is_primed(), "[{}] expected primed", i);

            let got = round_to(v, digits);
            let exp = round_to(expected[i], digits);
            assert_eq!(got, exp, "[{}] expected {}, got {}", i, exp, got);
        }
    }

    #[test]
    fn test_advance_decline_spot_checks() {
        let digits = 2;
        let highs = testdata::test_highs();
        let lows = testdata::test_lows();
        let closes = testdata::test_closes();
        let volumes = testdata::test_volumes();
        let count = highs.len();

        let mut ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();

        let mut values = Vec::new();
        for i in 0..count {
            let v = ad.update_hlcv(highs[i], lows[i], closes[i], volumes[i]);
            values.push(v);
        }

        let spot_checks: Vec<(usize, f64)> = vec![
            (0, -1631000.00),
            (1, 2974412.02),
            (250, 8707691.07),
            (251, 8328944.54),
        ];

        for (index, expected) in spot_checks {
            let got = round_to(values[index], digits);
            let exp = round_to(expected, digits);
            assert_eq!(got, exp, "spot check [{}]: expected {}, got {}", index, exp, got);
        }
    }

    #[test]
    fn test_advance_decline_update_bar() {
        let digits = 2;
        let mut ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();

        let highs = testdata::test_highs();
        let lows = testdata::test_lows();
        let closes = testdata::test_closes();
        let volumes = testdata::test_volumes();
        let expected = testdata::test_expected_ad();

        for i in 0..10 {
            let bar = Bar {
                time: 1_000_000,
                open: highs[i],
                high: highs[i],
                low: lows[i],
                close: closes[i],
                volume: volumes[i],
            };

            let output = ad.update_bar(&bar);
            let scalar = output[0].downcast_ref::<Scalar>().unwrap();

            let got = round_to(scalar.value, digits);
            let exp = round_to(expected[i], digits);
            assert_eq!(got, exp, "[{}] bar: expected {}, got {}", i, exp, got);
        }
    }

    #[test]
    fn test_advance_decline_scalar_update() {
        let mut ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();

        let v = ad.update(100.0);
        assert_eq!(v, 0.0, "expected 0 after scalar update, got {}", v);
        assert!(ad.is_primed(), "expected primed after first update");
    }

    #[test]
    fn test_advance_decline_nan() {
        let mut ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();

        assert!(ad.update(f64::NAN).is_nan());
        assert!(ad.update_hlcv(f64::NAN, 1.0, 2.0, 3.0).is_nan());
        assert!(ad.update_hlcv(1.0, f64::NAN, 2.0, 3.0).is_nan());
        assert!(ad.update_hlcv(1.0, 2.0, f64::NAN, 3.0).is_nan());
        assert!(ad.update_hlcv(1.0, 2.0, 3.0, f64::NAN).is_nan());
    }

    #[test]
    fn test_advance_decline_not_primed_initially() {
        let mut ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();

        assert!(!ad.is_primed());
        assert!(ad.update(f64::NAN).is_nan());
        assert!(!ad.is_primed());
    }

    #[test]
    fn test_advance_decline_metadata() {
        let ad = AdvanceDecline::new(&AdvanceDeclineParams).unwrap();
        let meta = ad.metadata();
        assert_eq!(meta.identifier, Identifier::AdvanceDecline);
        assert_eq!(meta.mnemonic, "ad");
        assert_eq!(meta.description, "Advance-Decline");
        assert_eq!(meta.outputs.len(), 1);
        assert_eq!(meta.outputs[0].mnemonic, "ad");
        assert_eq!(meta.outputs[0].description, "Advance-Decline");
    }
}
